### PR TITLE
Sema: Diagnose completely unapplied references to mutating methods.

### PR DESCRIPTION
**Pull request by jckarter**
The currying behavior of method references completely breaks in the face of `inout` semantics, even moreso with exclusivity enforcement, but we failed to diagnose these references in Swift 4 and previous versions. Raise a compatibility warning when these references are found in Swift 4 code, or error in Swift 5 and later. Simplify the partial application logic here slightly too now that standalone functions do not allow currying. Addresses rdar://problem/41361334 | SR-8074.

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@jckarter Since we always miscompile to undefined behavior, a warning sounds dangerous. Why not error out unconditionally?

**Review by slavapestov — APPROVED** [review]
LGTM, but perhaps you and the rest of the core team should decide to always make this an error instead.

**Comment by jckarter** [comment]
It feels a little late to spring new hard source breaks on users in 4.2 to me. I could ask though.

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@jckarter Fair enough.

**Comment by jckarter** [comment]
The core team agreed that a compatibility warning is the right approach here.

**Comment by jckarter** [comment]
@brentdax This patch didn't do anything to change `getStorageKind()`, could you be more specific?

**Comment by brentdax** [comment]
@jckarter You're right, I'm sorry. I must have mixed up tabs while I was tracing it down.

**Pull request merged**
